File: frontend/src/utils/pipeline.js
function parseDatetime(val) {
  if (!val) return null;
  const d = new Date(val);
  return isNaN(d.getTime()) ? null : d;
}

function isNumericCol(records, col) {
  let hasValues = false;
  for (const r of records) {
    const val = r[col];
    if (val !== undefined && val !== null && String(val).trim() !== "") {
      if (isNaN(Number(val))) {
        return false;
      }
      hasValues = true;
    }
  }
  return hasValues;
}

function isDatetimeCol(records, col) {
  const colLower = col.toLowerCase();
  if (!['time', 'date', 'created', 'updated', 'logged'].some(k => colLower.includes(k))) {
    return false;
  }
  
  let hasValues = false;
  for (const r of records) {
    const val = r[col];
    if (val !== undefined && val !== null && String(val).trim() !== "") {
      const dt = parseDatetime(val);
      if (dt !== null) {
        hasValues = true;
      } else {
        return false;
      }
    }
  }
  return hasValues;
}

function calculateMedian(values) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const n = sorted.length;
  if (n % 2 === 1) {
    return sorted[Math.floor(n / 2)];
  } else {
    return (sorted[n / 2 - 1] + sorted[n / 2]) / 2;
  }
}

function calculateMode(values) {
  if (values.length === 0) return 'Unknown';
  const counts = {};
  let maxCount = 0;
  let modeVal = 'Unknown';
  for (const v of values) {
    counts[v] = (counts[v] || 0) + 1;
    if (counts[v] > maxCount) {
      maxCount = counts[v];
      modeVal = v;
    }
  }
  return modeVal;
}

export function cleanAndProcessData(recordsInput) {
  if (!recordsInput || recordsInput.length === 0) {
    return {
      cleaned: [],
      report: {
        stats: {
          before: { row_count: 0, column_count: 0, duplicates: 0, nulls: {} },
          after: { row_count: 0, column_count: 0, duplicates: 0, nulls: {} }
        },
        data: [],
        anomalies: [],
        columns: []
      }
    };
  }

  // Deep copy
  let records = recordsInput.map(r => ({ ...r }));
  const totalRowsBefore = records.length;
  const columns = Object.keys(records[0]);

  // Deduplication
  const seen = new Set();
  const dedupedRecords = [];
  let duplicatesCount = 0;
  
  for (const r of records) {
    const sortedKeysValues = Object.keys(r).sort().map(k => `${k}:${r[k]}`).join('|');
    if (seen.has(sortedKeysValues)) {
      duplicatesCount++;
    } else {
      seen.add(sortedKeysValues);
      dedupedRecords.push(r);
    }
  }
  records = dedupedRecords;

  // Initial null counts & clean HTML/whitespace
  const nullsBefore = {};
  const dtypesBefore = {};
  const colTypes = {};
  
  for (const col of columns) {
    nullsBefore[col] = 0;
    dtypesBefore[col] = 'object';
  }

  for (const r of records) {
    for (const col of columns) {
      let val = r[col];
      if (val === undefined || val === null || String(val).trim() === "" || String(val).toLowerCase() === "nan" || String(val).toLowerCase() === "null" || String(val).toLowerCase() === "none") {
        r[col] = null;
        nullsBefore[col]++;
      } else {
        // Simple HTML unescaping
        let cleanedVal = String(val)
          .replace(/&amp;/g, '&')
          .replace(/&lt;/g, '<')
          .replace(/&gt;/g, '>')
          .replace(/&quot;/g, '"')
          .replace(/&#039;/g, "'")
          .trim();
        if (cleanedVal === "" || cleanedVal.toLowerCase() === "nan" || cleanedVal.toLowerCase() === "null" || cleanedVal.toLowerCase() === "none") {
          r[col] = null;
          nullsBefore[col]++;
        } else {
          r[col] = cleanedVal;
        }
      }
    }
  }

  // Detect column types
  for (const col of columns) {
    if (isNumericCol(records, col)) {
      colTypes[col] = 'numeric';
      dtypesBefore[col] = 'float64';
    } else if (isDatetimeCol(records, col)) {
      colTypes[col] = 'datetime';
      dtypesBefore[col] = 'datetime64[ns]';
    } else {
      colTypes[col] = 'categorical';
      dtypesBefore[col] = 'object';
    }
  }

  // Impute missing values
  for (const col of columns) {
    const type = colTypes[col];
    
    if (type === 'numeric') {
      const numericVals = [];
      for (const r of records) {
        if (r[col] !== null) {
          const num = Number(r[col]);
          if (!isNaN(num)) numericVals.push(num);
        }
      }
      const medianVal = calculateMedian(numericVals);
      
      for (const r of records) {
        if (r[col] === null) {
          r[col] = medianVal;
        } else {
          r[col] = Number(r[col]);
        }
      }
    } else if (type === 'datetime') {
      // Ffill
      let lastValid = null;
      for (const r of records) {
        if (r[col] !== null) {
          const dt = parseDatetime(r[col]);
          if (dt) lastValid = dt.toISOString().replace('T', ' ').substring(0, 19);
        }
        if (r[col] === null && lastValid !== null) {
          r[col] = lastValid;
        }
      }
      // Bfill
      lastValid = null;
      for (let i = records.length - 1; i >= 0; i--) {
        const r = records[i];
        if (r[col] !== null) {
          const dt = parseDatetime(r[col]);
          if (dt) lastValid = dt.toISOString().replace('T', ' ').substring(0, 19);
        }
        if (r[col] === null && lastValid !== null) {
          r[col] = lastValid;
        }
      }
      // Default fallback
      const fallbackNow = new Date().toISOString().replace('T', ' ').substring(0, 19);
      for (const r of records) {
        if (r[col] === null) {
          r[col] = fallbackNow;
        } else {
          const dt = parseDatetime(r[col]);
          if (dt) r[col] = dt.toISOString().replace('T', ' ').substring(0, 19);
        }
      }
    } else { // categorical
      const catVals = [];
      for (const r of records) {
        if (r[col] !== null) catVals.push(String(r[col]));
      }
      const modeVal = calculateMode(catVals);
      for (const r of records) {
        if (r[col] === null) {
          r[col] = modeVal;
        }
      }
    }
  }

  // Anomaly checks
  for (const r of records) {
    r.is_anomaly = false;
    r.anomaly_reason = "";
  }

  const anomalies = [];
  const numericCols = columns.filter(col => colTypes[col] === 'numeric' && col !== 'rating');

  if (numericCols.length > 0 && records.length > 0) {
    const n = records.length;
    const d = numericCols.length;
    
    // Build feature matrix X
    const X = [];
    for (let i = 0; i < n; i++) {
      const rowVals = [];
      for (let j = 0; j < d; j++) {
        const col = numericCols[j];
        const val = Number(records[i][col]);
        rowVals.push(isNaN(val) ? 0.0 : val);
      }
      X.push(rowVals);
    }

    // Standardize features (Z-score Scaling)
    const means = [];
    const stdDevs = [];
    for (let j = 0; j < d; j++) {
      const colVals = X.map(row => row[j]);
      const mean = colVals.reduce((a, b) => a + b, 0) / n;
      const variance = colVals.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / (n > 1 ? n - 1 : 1);
      const stdDev = Math.sqrt(variance);
      means.push(mean);
      stdDevs.push(stdDev);
    }

    const X_scaled = [];
    for (let i = 0; i < n; i++) {
      const scaledRow = [];
      for (let j = 0; j < d; j++) {
        const val = X[i][j];
        const mean = means[j];
        const stdDev = stdDevs[j];
        scaledRow.push(stdDev > 0 ? (val - mean) / stdDev : 0.0);
      }
      X_scaled.push(scaledRow);
    }

    // Method A: Z-score detection flags
    const zFlags = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < d; j++) {
        if (Math.abs(X_scaled[i][j]) > 3.0) {
          zFlags[i] = 1;
          break;
        }
      }
    }

    // Method B: Isolation Forest flags (contamination = 7%)
    const isoFlags = runIsolationForest(X_scaled, 100, 0.07);

    // Method C: DBSCAN flags (eps = 0.8, minSamples = 8)
    const dbFlags = runDbscan(X_scaled, 0.8, 8);

    // Ensemble Voting (agreed by >= 2 methods)
    for (let i = 0; i < n; i++) {
      const votes = zFlags[i] + isoFlags[i] + dbFlags[i];
      if (votes >= 2) {
        records[i].is_anomaly = true;
        const methods = [];
        if (zFlags[i] === 1) methods.push("Z-score");
        if (isoFlags[i] === 1) methods.push("Isolation Forest");
        if (dbFlags[i] === 1) methods.push("DBSCAN");
        records[i].anomaly_reason = `Ensemble outlier confirmed by ${votes}/3 methods (${methods.join(', ')}).`;
      }
    }
  }

  // Explicit rating bound check
  if (columns.includes('rating')) {
    for (let idx = 0; idx < records.length; idx++) {
      const r = records[idx];
      const val = r.rating;
      if (val !== null) {
        const fVal = Number(val);
        if (isNaN(fVal) || fVal < 1.0 || fVal > 5.0) {
          r.is_anomaly = true;
          const reason = `Out-of-bounds rating: ${val} (expected 1 to 5)`;
          r.anomaly_reason = r.anomaly_reason ? `${r.anomaly_reason} | ${reason}` : reason;
        }
      }
    }
  }

  // Extract anomalies
  for (let idx = 0; idx < records.length; idx++) {
    const r = records[idx];
    if (r.is_anomaly) {
      anomalies.push({
        row_index: idx,
        data: { ...r },
        reason: r.anomaly_reason
      });
    }
  }

  const report = {
    stats: {
      before: {
        row_count: totalRowsBefore,
        column_count: columns.length,
        duplicates: duplicatesCount,
        nulls: nullsBefore,
        dtypes: dtypesBefore
      },
      after: {
        row_count: records.length,
        column_count: columns.length,
        duplicates: 0,
        nulls: columns.reduce((acc, col) => ({ ...acc, [col]: 0 }), {}),
        dtypes: columns.reduce((acc, col) => ({
          ...acc,
          [col]: colTypes[col] === 'numeric' ? 'float64' : (colTypes[col] === 'datetime' ? 'datetime64[ns]' : 'object')
        }), {})
      },
      anomaly_count: anomalies.length
    },
    data: records,
    anomalies: anomalies,
    columns: columns.map(col => ({
      name: col,
      type: colTypes[col] === 'numeric' ? 'float64' : (colTypes[col] === 'datetime' ? 'datetime64[ns]' : 'object'),
      is_numerical: colTypes[col] === 'numeric',
      is_datetime: colTypes[col] === 'datetime',
      is_anomaly_candidate: colTypes[col] === 'numeric' && col !== 'rating'
    }))
  };

  return { cleaned: records, report };
}

// ============================================================================
// ENSEMBLE ANOMALY DETECTORS (PURE JAVASCRIPT IMPLEMENTATIONS)
// ============================================================================

function c_factor(n) {
  if (n <= 1) return 0;
  if (n === 2) return 1;
  return 2 * (Math.log(n - 1) + 0.5772156649) - (2 * (n - 1)) / n;
}

class IsolationTree {
  constructor() {
    this.splitFeature = null;
    this.splitValue = null;
    this.left = null;
    this.right = null;
    this.size = 0;
  }

  fit(X, depth, maxDepth) {
    this.size = X.length;
    if (X.length <= 1 || depth >= maxDepth) {
      return;
    }

    const numFeatures = X[0].length;
    const fIdx = Math.floor(Math.random() * numFeatures);
    this.splitFeature = fIdx;

    let min = X[0][fIdx];
    let max = X[0][fIdx];
    for (let i = 1; i < X.length; i++) {
      if (X[i][fIdx] < min) min = X[i][fIdx];
      if (X[i][fIdx] > max) max = X[i][fIdx];
    }

    if (min === max) {
      return;
    }

    const splitVal = min + Math.random() * (max - min);
    this.splitValue = splitVal;

    const leftX = [];
    const rightX = [];
    for (let i = 0; i < X.length; i++) {
      if (X[i][fIdx] < splitVal) {
        leftX.push(X[i]);
      } else {
        rightX.push(X[i]);
      }
    }

    this.left = new IsolationTree();
    this.left.fit(leftX, depth + 1, maxDepth);

    this.right = new IsolationTree();
    this.right.fit(rightX, depth + 1, maxDepth);
  }

  pathLength(x, currentDepth) {
    if (this.left === null || this.right === null) {
      return currentDepth + c_factor(this.size);
    }
    if (x[this.splitFeature] < this.splitValue) {
      return this.left.pathLength(x, currentDepth + 1);
    } else {
      return this.right.pathLength(x, currentDepth + 1);
    }
  }
}

function runDbscan(X_scaled, eps = 0.8, minSamples = 8) {
  const n = X_scaled.length;
  if (n === 0) return new Array(0).fill(0);
  const labels = new Array(n).fill(0); // 0 = unvisited
  let clusterId = 0;

  function getNeighbors(i) {
    const neighbors = [];
    const pi = X_scaled[i];
    for (let j = 0; j < n; j++) {
      const pj = X_scaled[j];
      let distSq = 0;
      for (let d = 0; d < pi.length; d++) {
        distSq += Math.pow(pi[d] - pj[d], 2);
      }
      if (Math.sqrt(distSq) <= eps) {
        neighbors.push(j);
      }
    }
    return neighbors;
  }

  function expandCluster(i, neighbors, clusterId) {
    labels[i] = clusterId;
    const queue = [...neighbors];
    for (let q = 0; q < queue.length; q++) {
      const neighborIdx = queue[q];
      if (labels[neighborIdx] === -1) {
        labels[neighborIdx] = clusterId;
      }
      if (labels[neighborIdx] !== 0) continue;

      labels[neighborIdx] = clusterId;
      const nextNeighbors = getNeighbors(neighborIdx);
      if (nextNeighbors.length >= minSamples) {
        for (let nn = 0; nn < nextNeighbors.length; nn++) {
          if (!queue.includes(nextNeighbors[nn])) {
            queue.push(nextNeighbors[nn]);
          }
        }
      }
    }
  }

  for (let i = 0; i < n; i++) {
    if (labels[i] !== 0) continue;

    const neighbors = getNeighbors(i);
    if (neighbors.length < minSamples) {
      labels[i] = -1; // noise
    } else {
      clusterId++;
      expandCluster(i, neighbors, clusterId);
    }
  }

  return labels.map(l => (l === -1 ? 1 : 0));
}

function runIsolationForest(X_scaled, nEstimators = 100, contamination = 0.07) {
  const n = X_scaled.length;
  if (n === 0) return new Array(0).fill(0);
  
  const sampleSize = Math.min(256, n);
  const maxDepth = Math.ceil(Math.log2(Math.max(2, sampleSize)));
  const trees = [];

  for (let i = 0; i < nEstimators; i++) {
    const sample = [];
    const indices = new Set();
    while (indices.size < sampleSize) {
      indices.add(Math.floor(Math.random() * n));
    }
    for (const idx of indices) {
      sample.push(X_scaled[idx]);
    }

    const tree = new IsolationTree();
    tree.fit(sample, 0, maxDepth);
    trees.push(tree);
  }

  const scores = [];
  const cVal = c_factor(sampleSize);
  
  for (let i = 0; i < n; i++) {
    let totalPathLength = 0;
    for (let t = 0; t < nEstimators; t++) {
      totalPathLength += trees[t].pathLength(X_scaled[i], 0);
    }
    const avgPathLength = totalPathLength / nEstimators;
    const score = cVal > 0 ? Math.pow(2, -avgPathLength / cVal) : 0.5;
    scores.push({ idx: i, score });
  }

  const sorted = [...scores].sort((a, b) => b.score - a.score);
  const cutoffIdx = Math.floor(n * contamination);
  const flaggedIndices = new Set(sorted.slice(0, cutoffIdx).map(item => item.idx));

  const flags = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    if (flaggedIndices.has(i)) {
      flags[i] = 1;
    }
  }
  return flags;
}

File: backend/pipeline.py
import csv
import json
import math
import html
import datetime
from collections import Counter

def parse_datetime(val):
    """
    Attempt to parse a datetime string. Returns datetime object or None.
    """
    if not val:
        return None
    if isinstance(val, (datetime.datetime, datetime.date)):
        return val
    val_str = str(val).strip()
    for fmt in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%dT%H:%M:%S", "%Y-%m-%d", "%Y/%m/%d %H:%M:%S", "%Y/%m/%d"):
        try:
            return datetime.datetime.strptime(val_str, fmt)
        except ValueError:
            continue
    return None

def is_numeric_col(records, col):
    """
    Check if a column is numeric (all non-null values can be converted to float).
    """
    has_values = False
    for r in records:
        val = r.get(col)
        if val is not None and val != "":
            # Check if it looks like float
            try:
                float(val)
                has_values = True
            except ValueError:
                return False
    return has_values

def is_datetime_col(records, col):
    """
    Check if a column name suggests datetime and contains parseable datetime strings.
    """
    col_lower = col.lower()
    if not any(k in col_lower for k in ['time', 'date', 'created', 'updated', 'logged']):
        return False
        
    has_values = False
    for r in records:
        val = r.get(col)
        if val is not None and val != "":
            dt = parse_datetime(val)
            if dt is not None:
                has_values = True
            else:
                return False
    return has_values

def calculate_median(values):
    """
    Calculate median of a numeric list in pure Python.
    """
    if not values:
        return 0
    sorted_vals = sorted(values)
    n = len(sorted_vals)
    if n % 2 == 1:
        return sorted_vals[n // 2]
    else:
        return (sorted_vals[n // 2 - 1] + sorted_vals[n // 2]) / 2.0

def clean_and_process_data(filepath_or_data):
    """
    Cleans, imputes, and runs anomaly detection on a dataset in pure Python.
    Returns: (cleaned_records, report_dict)
    """
    # 1. Parse input data into a list of dicts
    records = []
    if isinstance(filepath_or_data, str):
        # Determine if JSON or CSV
        striped_input = filepath_or_data.strip()
        if striped_input.startswith('[') or striped_input.startswith('{'):
            try:
                parsed = json.loads(striped_input)
                records = parsed if isinstance(parsed, list) else [parsed]
            except Exception:
                pass
        else:
            # Check if file path
            if os.path.exists(filepath_or_data):
                with open(filepath_or_data, 'r', encoding='utf-8-sig') as f:
                    content = f.read().strip()
                    if content.startswith('[') or content.startswith('{'):
                        records = json.loads(content)
                    else:
                        f.seek(0)
                        reader = csv.DictReader(f)
                        records = [dict(row) for row in reader]
            else:
                # Treat as CSV string
                reader = csv.DictReader(filepath_or_data.splitlines())
                records = [dict(row) for row in reader]
    elif isinstance(filepath_or_data, list):
        # Deep copy list of dicts
        records = [dict(row) for row in filepath_or_data]
    else:
        raise ValueError("Invalid input format for data processing.")

    total_rows_before = len(records)
    if total_rows_before == 0:
        return [], {
            'stats': {
                'before': {'row_count': 0, 'column_count': 0, 'duplicates': 0, 'nulls': {}},
                'after': {'row_count': 0, 'column_count': 0, 'duplicates': 0, 'nulls': {}}
            },
            'data': [],
            'anomalies': [],
            'columns': []
        }

    columns = list(records[0].keys())

    # Calculate duplicate count
    seen = set()
    deduped_records = []
    duplicates_count = 0
    for r in records:
        row_tuple = tuple((k, r[k]) for k in sorted(r.keys()))
        if row_tuple in seen:
            duplicates_count += 1
        else:
            seen.add(row_tuple)
            deduped_records.append(r)
            
    records = deduped_records

    # Capture initial nulls & types count
    nulls_before = {col: 0 for col in columns}
    dtypes_before = {col: 'object' for col in columns}
    
    # Pre-clean string values (unescape HTML, strip spaces)
    for r in records:
        for col in columns:
            val = r.get(col)
            if val is None or str(val).strip() == "" or str(val).lower() in ("nan", "null", "none"):
                r[col] = None
                nulls_before[col] += 1
            else:
                cleaned_val = html.unescape(str(val)).strip()
                if cleaned_val == "" or cleaned_val.lower() in ("nan", "null", "none"):
                    r[col] = None
                    nulls_before[col] += 1
                else:
                    r[col] = cleaned_val

    # Detect column types
    col_types = {}
    for col in columns:
        if is_numeric_col(records, col):
            col_types[col] = 'numeric'
            dtypes_before[col] = 'float64'
        elif is_datetime_col(records, col):
            col_types[col] = 'datetime'
            dtypes_before[col] = 'datetime64[ns]'
        else:
            col_types[col] = 'categorical'
            dtypes_before[col] = 'object'

    # Impute missing values
    for col in columns:
        col_type = col_types[col]
        
        if col_type == 'numeric':
            # Gather numerical values
            numeric_vals = []
            for r in records:
                if r[col] is not None:
                    try:
                        numeric_vals.append(float(r[col]))
                    except ValueError:
                        pass
            median_val = calculate_median(numeric_vals)
            
            # Fill nulls with median
            for r in records:
                if r[col] is None:
                    r[col] = median_val
                else:
                    r[col] = float(r[col])
                    
        elif col_type == 'datetime':
            # Run forward-fill and backward-fill
            last_valid = None
            # Ffill
            for r in records:
                val = r[col]
                if val is not None:
                    dt = parse_datetime(val)
                    if dt:
                        last_valid = dt.strftime('%Y-%m-%d %H:%M:%S')
                if val is None and last_valid is not None:
                    r[col] = last_valid
            
            # Bfill
            last_valid = None
            for r in reversed(records):
                val = r[col]
                if val is not None:
                    dt = parse_datetime(val)
                    if dt:
                        last_valid = dt.strftime('%Y-%m-%d %H:%M:%S')
                if val is None and last_valid is not None:
                    r[col] = last_valid
                    
            # Fallback to now
            now_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            for r in records:
                if r[col] is None:
                    r[col] = now_str
                else:
                    dt = parse_datetime(r[col])
                    if dt:
                        r[col] = dt.strftime('%Y-%m-%d %H:%M:%S')
                        
        else: # categorical
            # Gather mode
            non_null_cats = [r[col] for r in records if r[col] is not None]
            if non_null_cats:
                mode_val = Counter(non_null_cats).most_common(1)[0][0]
            else:
                mode_val = 'Unknown'
                
            for r in records:
                if r[col] is None:
                    r[col] = mode_val

    # Anomaly Detection
    for r in records:
        r['is_anomaly'] = False
        r['anomaly_reason'] = ""

    anomalies = []
    
    # 1. Ensemble anomaly detection for numerical columns (excluding rating)
    numeric_cols = [col for col, t in col_types.items() if t == 'numeric' and col != 'rating']
    
    if numeric_cols and len(records) > 0:
        import numpy as np
        from sklearn.preprocessing import StandardScaler
        from sklearn.ensemble import IsolationForest
        from sklearn.cluster import DBSCAN
        from scipy import stats
        
        # Build X matrix
        X = []
        for r in records:
            row_vals = []
            for col in numeric_cols:
                try:
                    row_vals.append(float(r[col]))
                except ValueError:
                    row_vals.append(0.0)
            X.append(row_vals)
        X = np.array(X)
        
        # Preprocessing: Standard Scaler
        scaler = StandardScaler()
        X_scaled = scaler.fit_transform(X)
        
        # Method A: Z-score
        z = np.abs(stats.zscore(X_scaled, axis=0))
        z = np.nan_to_num(z, nan=0.0)
        z_flags = (z > 3.0).any(axis=1).astype(int)
        
        # Method B: Isolation Forest (contamination = 7%)
        iso = IsolationForest(
            n_estimators=100,
            contamination=0.07,
            max_samples="auto",
            random_state=42
        )
        iso_preds = iso.fit_predict(X_scaled)
        iso_flags = (iso_preds == -1).astype(int)
        
        # Method C: DBSCAN (eps = 0.8, min_samples = 8)
        db = DBSCAN(eps=0.8, min_samples=8)
        db_labels = db.fit_predict(X_scaled)
        db_flags = (db_labels == -1).astype(int)
        
        # Ensemble Voting (agreed by >= 2 methods)
        votes = z_flags + iso_flags + db_flags
        confirmed_flags = (votes >= 2).astype(int)
        
        for idx, r in enumerate(records):
            v = votes[idx]
            if confirmed_flags[idx] == 1:
                r['is_anomaly'] = True
                methods = []
                if z_flags[idx] == 1: methods.append("Z-score")
                if iso_flags[idx] == 1: methods.append("Isolation Forest")
                if db_flags[idx] == 1: methods.append("DBSCAN")
                r['anomaly_reason'] = f"Ensemble outlier confirmed by {v}/3 methods ({', '.join(methods)})."

    # 2. Bounded rule check for rating
    if 'rating' in columns:
        for idx, r in enumerate(records):
            val = r.get('rating')
            if val is not None:
                try:
                    f_val = float(val)
                    if f_val < 1.0 or f_val > 5.0:
                        r['is_anomaly'] = True
                        reason = f"Out-of-bounds rating: {f_val} (expected 1 to 5)"
                        if r['anomaly_reason']:
                            r['anomaly_reason'] += f" | {reason}"
                        else:
                            r['anomaly_reason'] = reason
                except ValueError:
                    r['is_anomaly'] = True
                    reason = f"Invalid rating format: {val}"
                    if r['anomaly_reason']:
                        r['anomaly_reason'] += f" | {reason}"
                    else:
                        r['anomaly_reason'] = reason

    # Package findings
    for idx, r in enumerate(records):
        if r['is_anomaly']:
            anomalies.append({
                'row_index': idx,
                'data': r,
                'reason': r['anomaly_reason']
            })

    # Prepare report metadata
    report = {}
    report['before'] = {
        'row_count': total_rows_before,
        'column_count': len(columns),
        'duplicates': duplicates_count,
        'nulls': nulls_before,
        'dtypes': dtypes_before
    }
    
    report['after'] = {
        'row_count': len(records),
        'column_count': len(columns),
        'duplicates': 0,
        'nulls': {col: 0 for col in columns},
        'dtypes': {col: 'float64' if col_types[col] == 'numeric' else ('datetime64[ns]' if col_types[col] == 'datetime' else 'object') for col in columns}
    }
    
    report['anomaly_count'] = len(anomalies)

    cols_meta = []
    for col in columns:
        cols_meta.append({
            'name': col,
            'type': 'float64' if col_types[col] == 'numeric' else ('datetime64[ns]' if col_types[col] == 'datetime' else 'object'),
            'is_numerical': bool(col_types[col] == 'numeric'),
            'is_datetime': bool(col_types[col] == 'datetime'),
            'is_anomaly_candidate': bool(col_types[col] == 'numeric' and col != 'rating')
        })

    return records, {
        'stats': report,
        'data': records,
        'anomalies': anomalies,
        'columns': cols_meta
    }

# Import guard
import os
